Refactoring

leoisl opened the pull request:
This is a refactoring of the GCP library to some standards I normally use (thus it is somewhat personal, feel free to reject if you don't like these standards). The logic is basically the same, but with lots of refactoring. Main changes are:

1. Separation between interface and implementation in the header file (main goal of the library is just its use. Implementation is detail. Right now we have more lines of implementation than interface/declaration). User is not interested in implementation details. If he/she is, he/she can scroll to the bottom to find the implementation details.

2. In if statements now we only have a descriptive boolean variable describing the test being done;

3. Fragmentation of long methods into shorter ones with single responsibility;

4. Some variable renaming to more descriptive naming;

PS: we have some previous commits (not in this PR, already merged to master) that: 1) sorts genotype confidences in the constructor; 2) adds `Percentiler::get_confidence_percentiles()`; 3) adds more small tests, and two huge tests. But I think these 3 changes did not need a review, like this one.

Commit:
Refactoring GCP

Commit:
Updating CMakeLists.txt

bricoletc commented on src/GCP.h:
```diff
@@ -101,97 +104,206 @@ class NotEnoughData : public std::runtime_error {
+}
+
+GenotypePercentile Percentiler::get_confidence_percentile(GenotypeConfidence queried_confidence) const {
+  auto upper_bound = confidence_to_percentile.upper_bound(queried_confidence);
```
Ive just realised this has to be a call to `lower_bound` not upper bound. The latter always gives the first entry greater than the query. So we never return a known value. Linear interpolation is always called, and still gives the right result. But this is a bug (i think i may have introduced it).

leoisl replied:
Really nice spotting! I refactored this function and did not see that. Fixed this here: https://github.com/leoisl/GCP/pull/10/commits/82bb64858e82477cd53f79ef4279db9467fbd528 (with two other smaller changes/fixes).

After these changes, one of the huge tests started failing. It turns out that I thought Minos had a bug, and I changed one of the answers to what I thought was correct. I actually got it wrong: Minos output was correct. This bugfix also fixed this previous bug, and now Minos output and our output are exactly the same for 4 tests (2 small and 2 huge tests) (commit fixing the tests: https://github.com/leoisl/GCP/pull/10/commits/41028b8c1a2ca86d8860abd817074c5751c7aaaa ).

bricoletc replied:
great!

bricoletc reviewed (changes requested):
There's one bug I caught reading this- apart from that looks fine to me!

Just wondering is the `inline` required somehow to allow separate declaration/definition in the same source file?

leoisl commented:
> Just wondering is the `inline` required somehow to allow separate declaration/definition in the same source file?

Yep! When separating the method definition from the declaration to the bottom of the file, I was getting an error saying that there was multiple definitions of the separated methods. I found it weird since we were actually defining the method only once (but at the bottom of the file), and the file is protected with the `#ifndef`, `#define` and `#endif` macros, so I did not actually understand. Some searching, and this explains: https://stackoverflow.com/a/14517625/5264075 . Summary is `inline` is implicitly set when defining the method inside the class, and avoids the multiple definition error. If we define the method outside of the class, we have then to explicitly state it is inline to avoid this error.

PS: after writing this, I knew that `inline` solved the problem, but I actually did not know why... Well the main use of code reviews is to learn from each other and improve code, and sometimes it is nice to understand the reasons, as C++ can be tricky!

I removed the `inline` from one of the methods, and the issue is actually a linker error that happens **only** when compiling the tests, not the examples. To understand the issue, read this answer: https://stackoverflow.com/a/14517957/5264075 which explains how the preprocessor produces `cpp` files for the compiler. The issue we get is:

```
[ 75%] Linking CXX executable test_main
cd /home/leandro/git/GCP/cmake-build-debug-local/tests && /home/leandro/Applications/CLion-2019.1.2/clion-2019.1.2/bin/cmake/linux/bin/cmake -E cmake_link_script CMakeFiles/test_main.dir/link.txt --verbose=1
/usr/bin/c++   -std=c++11 -Wall -Wextra -g  -rdynamic CMakeFiles/test_main.dir/test_percentiler.cpp.o CMakeFiles/test_main.dir/test_percentiler_big.cpp.o CMakeFiles/test_main.dir/utils_test.cpp.o  -o test_main  ../lib/libgtest_maind.a ../lib/libgtestd.a -lpthread 
CMakeFiles/test_main.dir/test_percentiler_big.cpp.o: In function `testing::Test::SetUpTestSuite()':
/home/leandro/git/GCP/src/GCP.h:303: multiple definition of `GCP::Percentiler::linear_interpolation(double, double, double, double, double)'
CMakeFiles/test_main.dir/test_percentiler.cpp.o:/home/leandro/git/GCP/src/GCP.h:303: first defined here
CMakeFiles/test_main.dir/utils_test.cpp.o: In function `__gnu_cxx::__ops::__iter_less_iter()':
/home/leandro/git/GCP/src/GCP.h:303: multiple definition of `GCP::Percentiler::linear_interpolation(double, double, double, double, double)'
CMakeFiles/test_main.dir/test_percentiler.cpp.o:/home/leandro/git/GCP/src/GCP.h:303: first defined here
collect2: error: ld returned 1 exit status
tests/CMakeFiles/test_main.dir/build.make:115: recipe for target 'tests/test_main' failed
make[2]: *** [tests/test_main] Error 1
```

This line shows the linker command:
```
/usr/bin/c++   -std=c++11 -Wall -Wextra -g  -rdynamic CMakeFiles/test_main.dir/test_percentiler.cpp.o CMakeFiles/test_main.dir/test_percentiler_big.cpp.o CMakeFiles/test_main.dir/utils_test.cpp.o  -o test_main  ../lib/libgtest_maind.a ../lib/libgtestd.a -lpthread 
```

The issue is that the three `.cpp` files we have (`test_percentiler.cpp`, `test_percentiler_big.cpp`, `utils_test.cpp`) will each include and contain `GCP.h`, which contains definitions of the methods. Compiler will successfully compile each of these `cpp` files separately, but then will try to merge them into a single executable by linking them. When doing the linking, we actually have 3 definitions for each non-inlined method. `inline` somehow avoids this.

To check if it is actually this, we can do some copy pasting, and avoid having 3 `cpp` files, putting everything in a single `cpp` file. In this case, there is no need for the `inline`, as the linker will see only one definition of each method. So, without the inline, we can't link different `cpp` files including `GCP.h`.

Commit:
Fixing Percentiler::get_confidence_percentile()

- Using lower_bound() instead of upper_bound() (issue spotted by Brice: https://github.com/leoisl/GCP/pull/10#discussion_r433701922)
- Using an double tolerance comparison instead of double equality;
- Min/max confidence check is done without the lower_bound iterator (a bit easier to read and implement)

Commit:
Fixing a bug in HugeTestWithDuplicates

All examples now match exactly Minos output

leoisl commented:
Fixed the bug you pointed out. Now, Minos' and this library's output on 4 tests (2 small, and 2 huge) are exactly the same, I think we have a correct implementation now.

bricoletc commented:
Yeah for the inline thanks for the explanation, im guessing the `.o` files each get a full copy of the function definition, which avoids breaking the one definition rule. Makes sense to me!

iqbal-lab commented:
When do we do double comparison @leoisl @bricoletc ?

leoisl commented:
> When do we do double comparison @leoisl @bricoletc ?

Here: https://github.com/leoisl/GCP/blob/41028b8c1a2ca86d8860abd817074c5751c7aaaa/src/GCP.h#L278 . I think classical double equality works in all our tests, but we could have this bug in production... Flaot/Double comparison seems tricky: https://stackoverflow.com/questions/17333/what-is-the-most-effective-way-for-float-and-double-comparison/
